File: neet_code_150_cpp/bit_manipulation/missing_number.cpp
#include <iostream>
#include <vector>

class SolutionAns {
public:
    int missingNumber(std::vector<int>& nums) {
        int n = nums.size();
        int result = n;
        for (int i = 0; i < n; i++) {
            result ^= i ^ nums[i];
        }
        return result;
    }
};

int main(void) {
    auto case_1 = std::vector{3, 0, 1};
    // => 2
    auto case_2 = std::vector{0, 1};
    // => 2
    auto case_3 = std::vector{9, 6, 4, 2, 3, 5, 7, 0, 1};
    // => 8

    SolutionAns s_ans;
    std::cout << s_ans.missingNumber(case_1) << std::endl;
    std::cout << s_ans.missingNumber(case_2) << std::endl;
    std::cout << s_ans.missingNumber(case_3) << std::endl;
}

File: neet_code_150_go/bit_manipulation/missing_number.go
package main

import "fmt"

// AC
func missingNumber(nums []int) int {
	result := 0
	for i, n := range nums {
		result ^= n
		result ^= (i + 1)
	}

	return result
}

func main() {
	case_1 := []int{3, 0, 1}
	// => 2
	case_2 := []int{0, 1}
	// => 2
	case_3 := []int{9, 6, 4, 2, 3, 5, 7, 0, 1}
	// => 8

	fmt.Println("case_1: ", missingNumber(case_1))
	fmt.Println("case_2: ", missingNumber(case_2))
	fmt.Println("case_3: ", missingNumber(case_3))
}
